### DISCUSSION TITLE
zigpy project roadmap for ZHA WebSocket Server ("ZHAWS") and matching ZHA WS addon + client integration and external/stand-alone ZHA Gateway WebSocket Server application?

### DISCUSSION THREAD

**Discussion by Hedda** [zigpy/zigpy#1207] [ZHA development]
Wondering if [ZHA integration](https://www.home-assistant.io/integrations/zha) and [zigpy](https://github.com/zigpy/zigpy) developers have plans to continue working on dmulcahey's ZHA WebSocket Server ("ZHAWS") with "zha-websocket-server", "zhaws-addon", and the matching "zha-ws fork" for client-side which I understand aims to breakout and create a stand-alone Zigbee gateway application from the `zha` component in Home Assistant's core so can run as a stand-alone Zigbee Gateway application that can be updated separatley from Home Assistant's core?

That is, are you still considering the idea of breaking out parts of the [ZHA component from the Home Assistant's core ](https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha) into an external library separate add-on and/or even stand-alone server software so it instead would use a [distributed application structure](https://en.wikipedia.org/wiki/Distributed_computing) that follows a [client–server model architecture](https://en.wikipedia.org/wiki/Client%E2%80%93server_model)? Or did I understand the concept wrong and/or misunderstand the goal? 

* https://github.com/zigpy/zha-websocket-server/
* https://github.com/zigpy/zhaws-addon
* https://github.com/dmulcahey/home-assistant/tree/dm/zha-ws
  * https://github.com/dmulcahey/home-assistant/tree/dm/zha-ws/homeassistant/components/zhaws

My guess was that the end result should then work similarly to the [Z-Wave JS add-on](https://www.home-assistant.io/integrations/zwave_js) which includes the Z-Wave JS Server to be a part of the Z-Wave JS UI integration?

* https://github.com/zwave-js/zwave-js-server
* https://github.com/home-assistant/addons/tree/master/zwave_js
  * https://github.com/home-assistant/addons/blob/master/zwave_js/DOCS.md
* https://github.com/home-assistant/core/tree/dev/homeassistant/components/zwave_js

I believe using a design similar to Z-Wave JS add-on could bring at least these benefits compared to ZHA's current monolithic design?

- Enabling out-of-band updates of ZHAWS add-on (including its dependencies), so must not follow the Home Assistant core releases. 
- Enabling running server instance on a different computer on the same local network (like for ex. on a Raspberry Pi Zero over Wi-Fi).
- Enable ZHA integration's IoT class to be listed as "Local Push" instead of "Local Polling" for direct real-time state feedback/notifications.
  - https://github.com/home-assistant/home-assistant.io/issues/26125
- Enabling running multiple server instances (with each server instance using its own Zigbee Coordinator for separate Zigbee PAN).
  - The ability to run multiple instances of ZHA inside Home Assistant, or rather the ability for ZHA to support several Zigbee Coordinator adapters with separate Zigbee networks, is an very often requested end-user question in the community from people who have areas far from their main house, like a garage or workshop.
- WebServer API should also make it simpler for third parties making third-party application implementations (not related to Home Assistant).
  - Enabling the ability to have multiple UI applications to control the zha-websocket-server. 
  - Someone else could also create other server wrappers or clients (like ex. Matter, MQTT, or NodeJS), as well as possibly make it usable as a stand-alone application as well.

References:

_Originally posted by [puddly](https://github.com/puddly) in https://github.com/zigpy/zigpy-cli/issues/7#issuecomment-1039521822_
> _You may be interested in [dmulcahey](https://github.com/dmulcahey)'s https://github.com/zigpy/zha-websocket-server/ (and ZHA WS itself https://github.com/dmulcahey/home-assistant/tree/dm/zha-ws) as a basis for such a tool.  Its command line client currently is just thin frontend for sending JSON but building an alternative CLI with asyncio-native [python-prompt-toolkit](https://python-prompt-toolkit.readthedocs.io/en/master/) seems very doable in the future._
> 
> _Once the low-level network information PRs are done, adding backup/restore to ZHA WS (and things from https://github.com/mdeweerd/zha-toolkit/!) will be very easy._

PS: FYI, Home Assistant itself by the way also offer a Home Assistant websocket API and a JavaScript websocket client as an example client:

https://github.com/home-assistant/home-assistant-js-websocket

**Comment by MattWestb** [comment]
Im finding more interesting making ZHA ZCL R9 / Zigbe R23 compatible and can handle over its devices to HAs Matter integration witch making its 100% future secure.
The problems is not zigpy then its very Zigbee but all special made in ZHA for strange devices like Xiaomi / Aqara and tuya that we cant getting ZCL compatible => not Matter compatible.
But most IKEA and other good behaving device can being used with out any problem and perhaps HA is getting Matter bridge functionality in the future then you can using the multi fabric function on all your Zigbee devices also if they is old but good ZCL compatible (like our famous bad OSRAM Plugs).

So dont forgetting that Matter is one evolution of Zigbee and is made being compatible and not losing functions and adding more advanced to devices without braking the compatibility as long flowing the specifications.

**Reply by Hedda** [reply]
That is all waaaaay off-topic from the original subject though and might just be the definition of the idiom to "[go off on a tangent](https://www.merriam-webster.com/dictionary/go%20off%20on%20a%20tangent)" 😄 

Matter does not matter here in ZHA (and zigpy) for all intents and purposes as Zigbee will still continue to live separately on its own.

**Comment by Hedda** [comment]
It looks like a ZHA and zigpy developer dmulcahey have now begun work splitting/breaking-out the the [ZHA Gateway](https://www.home-assistant.io/integrations/zha) part from Home Assistant ZHA component into an external stand-alone library with similar event based API:


* https://github.com/zigpy/zha
  * https://github.com/zigpy/zha/blob/dev/zha/application/gateway.py
* For a developers overview see https://github.com/zigpy/zha/pull/6 

And a new matching fork of a new ZHA integration  application component for Home Assistant broken out and using said stand-alone and external implementation of "ZHA Gateway":

* https://github.com/dmulcahey/home-assistant/tree/dm/zha-external-zha

As I understand it this initial external/stand-alone ”ZHA Gateway” project is very much a work-in-progress and that "*things will be broken for a while*" in it.

Still it looks like have come along far with this new concept, and looks initially to be a forked and restructured variant of Home Assistant's `zha` component that is converted into a standalone program:

[https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha](https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha)

For reference; the [ZHA integration](https://www.home-assistant.io/integrations/zha) depends on libraries from the [zigpy](https://github.com/zigpy/zigpy) project but the Zigbee gateway implementation that is the is [ZHA integration](https://www.home-assistant.io/integrations/zha) is today a built-into Home Assistant core, which means it can not be updated or maintained separately and it stops running when to stop or restart the Home Asistant core.

I guess the idea is still to split out the ZHA component from Home Assistant core has grown from a previous concept that the same developer(s) began working  on couple years ago with a zha-websocket-server and zha-ws client addon/plugin for Home Assistant?

[zigpy/zigpy#1207](https://github.com/zigpy/zigpy/discussions/1207)

[https://github.com/zigpy/zha-websocket-server](https://github.com/zigpy/zha-websocket-server)

[https://github.com/zigpy/zhaws-addon](https://github.com/zigpy/zhaws-addon)

[https://github.com/dmulcahey/home-assistant/tree/dm/zha-ws](https://github.com/dmulcahey/home-assistant/tree/dm/zha-ws)

[https://github.com/dmulcahey/home-assistant/tree/dm/zha-ws/homeassistant/components/zhaws](https://github.com/dmulcahey/home-assistant/tree/dm/zha-ws/homeassistant/components/zhaws)

I assume the goal will to breakout the `zha` component from Home Assistant core and make it an external Zigbee gateway application that can be updated separately from Home Assistant core and in theory also enable it to be used as a stand-alone and/or external version of ZHA as Zigbee gateway implementations by others in other third-party projects which then could make their own integrations and/or different UI frontends for that stand-alone version of zha, (probably accessed via websocket), or?

As splitting the Zigbee gateway application implemnentation part in the zha component away from the Home Assistant core should allow you to run "ZHA Gateway" externally, either as an addon or even stand-alone on a different computer, with it only communicating with the new ZHA integration in Home Assistant using websocket (which can be done over a local network). Today this is probably the most common argument why you might still want to recommend [Zigbee2MQT](https://www.zigbee2mqtt.io/)T instead of Home Assistant's built-in native [ZHA integration](https://www.home-assistant.io/integrations/zha) to a brand new Zigbee user (as Z2M and ZHA today otherwise compare equally in features and functions), or?

Another potential bonus for ZHA users is that it could potentially in the future also eanble users to have multiple instanses of external ZHA Gateway servers (each with its own Zigbee Coorinator radio adapter only controlling its own seperate Zigbee network), allowing users to for example have a "production Zigbee network" and a "test Zigbee network" (to seperate the testing of new and troublesome Zigbee devides). 

PS: If this case the result will again likely become very similar to the "Z-Wave JS" (node-zwave-js) project which uses a similar [client–server model](https://en.wikipedia.org/wiki/Client%E2%80%93server_model) (also known as "frontend-backend model") [distributed application](https://en.wikipedia.org/wiki/Distributed_application) structure architecture for its different implementations which depends on an addon (plugin) and external server program that can also be used as a stand-alone application. For reference check out the Z-Wave JS integration and related projects :

[https://www.home-assistant.io/integrations/zwave_js/](https://www.home-assistant.io/integrations/zwave_js/)

[https://github.com/home-assistant/addons/tree/master/zwave_js](https://github.com/home-assistant/addons/tree/master/zwave_js)

[https://github.com/zwave-js/node-zwave-js](https://github.com/zwave-js/node-zwave-js)

[https://github.com/zwave-js/zwave-js-server](https://github.com/zwave-js/zwave-js-server)

[https://github.com/zwave-js/zwave-js-ui](https://github.com/zwave-js/zwave-js-ui)

PS: The main zigpy library and the radio libraries will continue to be used by ZHA as low-level Zigbee and hardware-abratraction respectively.

**Comment by Hedda** [comment]
Any updates to original concept of the ZHA WebSocket Server (ZHA WS) project?

https://github.com/zigpy/zha-websocket-server

Asking since noticed that ZHA/zigpy and Home Assistant Core developers have now merged your pull request with changes that split/broken-out most of the ZHA component code from Home Assistant Core and move that into an external zha library:

https://github.com/home-assistant/core/pull/120190

Is that external ZHA library an initial part of and first step in breaking out ZHA with a goal to eventually be able to run ZHA as distributed or stand-alone application using a client-server model in the future?

https://github.com/zigpy/zha

Quoting puddly from the pull request linked above for Home Assistant Core:

> # _Proposed change_
> 
> _Today, ZHA is the largest integration in terms of lines of code (33% larger than the second-largest, MQTT)_
> 
> _Chart showing how ZHA dropping from the number one integration in terms of lines of code to about number ten_
> 
> _dmulcahey, TheJulianJES, and I have spent a long time migrating the core logic of ZHA into [a self-contained library](https://github.com/zigpy/zha). The end result is huge improvement in terms of code quality in both the Home Assistant half and the library half. **With this change (and others in the future), we aim to make ZHA reusable outside of Home Assistant, easier to understand, and pave the path to a more streamlined model for contributing device support code to ZHA and Home Assistant (when necessary)**._
> 
> _There should be no functional changes with this PR: everything is intended to work exactly as it did before._
> 
> _This is an initial pass and we aimed to modify as little as possible in this split. You will notice that the terminology currently in use within the library (and thus the rewritten integration) mirrors that of Core. This is intentional and will diverge as we see fit._
